Rando Enhancement: Mysterious Shuffled Items

Malkierian opened the pull request:
Adds a rando enhancement toggle which obfuscates shuffled freestanding/drawn-in-world items (PoH, tokens, shop items, keys, etc) with a custom question mark model so you never know what you're getting until you get it. Also utilizes `GetMerchantMessage`'s mysterious option directly based on the setting, so is applied to all item names in purchasable situations.

Color is editable with cosmetics editor under the World & NPCs tab, and can be rainbow.

Also hides item names in shops for items with the Seen/Identified status when the enhancement is on.

Adds `IsCheckShuffled` to allow for the distinction between seed settings and visibility options, which is used to determine whether to feed the fake `GetItemEntry` for the mystery model to `GetItemEntry_Draw`, which then calls the custom draw function associated with it.

Credit to @PurpleHato for the model.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1471612195.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1471618491.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1471619742.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1471626968.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/1471628169.zip)
<!--- section:artifacts:end -->

Malkierian commented:
https://github.com/HarbourMasters/Shipwright/assets/774771/5d65960e-9a0d-4f19-bc0c-66d824c5c779

jbodner09 commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -1537,6 +1537,10 @@ void DrawRandomizerMenu() {
             );
+            UIWidgets::PaddedEnhancementCheckbox("Freestanding Mystery Items", "gFreestandingMystery", true, false);
+            UIWidgets::Tooltip(
+                "Displays a \"Mystery Item\" model in place of any freestanding/GS/shop items that were shuffled,\n"
```
nit: Don't need to manually specify line breaks just for long lines

Malkierian replied:
Oh, right, I remembered seeing that tooltip in practice and it didn't look like it even did anything XD

Malkierian commented:
This is now ready for merge scrutiny.

I will mention, currently there's a bug in rando loading that holds onto `merchantPrices` across file/spoiler loading, so while this is setup to show/hide shuffled item names properly in the check tracker, that may not be displayed properly because you might have other saves with items shuffled that aren't shuffled on your current save that the check against `merchantPrices` would then find. This can be mitigated by deleting all other saves, generating a new seed, restarting SoH entirely, and then creating the new file without generating a different seed during the playthrough.

Malkierian commented:
OK, this is broken now because of the changes to OTRExporter and the model stuff I had going away. I don't think I can fix that until the SoH asset location PRs get merged, so in the meantime, if you want to test it, use the attached soh.otr to avoid crashes.
[soh.zip](https://github.com/HarbourMasters/Shipwright/files/13079009/soh.zip)

Malkierian committed:
Add rando enhancement "Mysterious Shuffled Items", which obfuscates shuffled freestanding/drawn-in-world items (PoH, tokens, shop items) with a custom question mark model (thanks Hato), and uses the "mysterious item" functionality of `GetMerchantMessage` for everything that supports it, regardless of hint status on generation.

Malkierian commented:
OK, this should be back to normal now. 🤞

Malkierian commented:
OK, now that I know the slowdown in main file select isn't my fault, this is fully ready to go.

Malkierian committed:
Reverted back to rando enhancement, but added condition for Mysterious Shuffle not being on for the gem rotation fix to apply.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into freestanding-mystery

Malkierian commented:
Guess I'm going to have to remake this for rando v3.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into freestanding-mystery

Malkierian committed:
First attempt at changing to fake GetItemEntry instead of directly calling the mystery draw function. Needs more work.

Malkierian committed:
Merge branch 'freestanding-mystery' of github.com:Malkierian/Shipwright into freestanding-mystery

Malkierian committed:
Updated CVar to reflect CVar rework values.

Added `IsCheckShuffled` as preliminary function for checking *only* if a check is shuffled, not necessarily if it is just visible on the tracker. This accounts for the difference between tokensanity and "Always Show GS On Tracker", where you don't want to obfuscate the latter.

Malkierian committed:
Bit of cleanup.

Malkierian commented:
Hopefully ready to go for the last time XD.

briaguya0 commented on soh/src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:
```diff
@@ -99,8 +99,8 @@ void ItemBHeart_Draw(Actor* thisx, PlayState* play) {
-        GetItemEntry_Draw(play, Randomizer_GetItemFromActor(this->actor.id, 
-            play->sceneNum,this->actor.params, GI_HEART_CONTAINER_2));
+        RandomizerCheck check = Randomizer_GetCheckFromActor(this->actor.id, play->sceneNum, this->actor.params);
+        GetItemEntry_Draw(play, (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(check)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(check, GI_HEART_CONTAINER_2), check);
```
removing the extra `check` at the end gets past the error that is visible in CI at the moment, but when trying to build locally after that change i started getting a ton of multiple definition errors (the "move to a function" thing makes those go away)

```suggestion
        GetItemEntry_Draw(play, (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(check)) ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(check, GI_HEART_CONTAINER_2));
```

briaguya0 commented on soh/soh/OTRGlobals.h:
```diff
@@ -165,6 +165,8 @@ GetItemEntry Randomizer_GetItemFromActor(s16 actorId, s16 sceneNum, s16 actorPar
 GetItemEntry Randomizer_GetItemFromKnownCheck(RandomizerCheck randomizerCheck, GetItemID ogId);
 GetItemEntry Randomizer_GetItemFromKnownCheckWithoutObtainabilityCheck(RandomizerCheck randomizerCheck, GetItemID ogId);
+bool Randomizer_IsCheckShuffled(RandomizerCheck check);
+GetItemEntry GetItemMystery;
```
```suggestion
GetItemEntry GetItemMystery();
```

turning this into a function got it to build on linux for me. i assume there's a better way to do this but this at least makes it work

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -137,6 +138,8 @@ Color_RGB8 kokiriColor = { 0x1E, 0x69, 0x1B };
 Color_RGB8 goronColor = { 0x64, 0x14, 0x00 };
 Color_RGB8 zoraColor = { 0x00, 0xEC, 0x64 };
 
+extern "C" GetItemEntry GetItemMystery = { ITEM_NONE_FE, 0, 0, 0, 0, 0, 0, 0, 0, false, ITEM_FROM_NPC, ITEM_CATEGORY_JUNK, NULL, MOD_RANDOMIZER, (CustomDrawFunc)Randomizer_DrawMysteryItem };
```
```suggestion
extern "C" GetItemEntry GetItemMystery() {
    return { ITEM_NONE_FE, 0, 0, 0, 0, 0, 0, 0, 0, false, ITEM_FROM_NPC, ITEM_CATEGORY_JUNK, NULL, MOD_RANDOMIZER, (CustomDrawFunc)Randomizer_DrawMysteryItem };
}
```

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -1367,8 +1367,8 @@ void EnItem00_DrawCollectible(EnItem00* this, PlayState* play) {
-            this->randoGiEntry =
-                Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
+            this->randoGiEntry = CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(randoCheck)
+                                     ? GetItemMystery : Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
```
```suggestion
                                     ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
```

briaguya0 commented on soh/src/code/z_en_item00.c:
```diff
@@ -1461,8 +1461,8 @@ void EnItem00_DrawHeartPiece(EnItem00* this, PlayState* play) {
-            this->randoGiEntry =
-                Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
+            this->randoGiEntry = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(randoCheck))
+                                    ? GetItemMystery : Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
```
```suggestion
                                    ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(randoCheck, GI_NONE);
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c:
```diff
@@ -515,14 +515,17 @@ void EnExItem_DrawItems(EnExItem* this, PlayState* play) {
             case EXITEM_BOMB_BAG_COUNTER:
-                randoGetItem = Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE, GI_BOMB_BAG_20);
+                randoGetItem = (CVarGetInteger("gEnhancement.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE))
+                                ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE, GI_BOMB_BAG_20);
```
```suggestion
                                ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE, GI_BOMB_BAG_20);
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c:
```diff
@@ -515,14 +515,17 @@ void EnExItem_DrawItems(EnExItem* this, PlayState* play) {
             case EXITEM_BOMBCHUS_COUNTER:
-                randoGetItem = Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_BOMBCHUS, GI_BOMBCHUS_10);
+                randoGetItem = (CVarGetInteger("gEnhancement.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_MARKET_BOMBCHU_BOWLING_BOMBCHUS))
+                                ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_BOMBCHUS, GI_BOMBCHUS_10);
```
```suggestion
                                ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_BOMBCHUS, GI_BOMBCHUS_10);
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c:
```diff
@@ -515,14 +515,17 @@ void EnExItem_DrawItems(EnExItem* this, PlayState* play) {
             case EXITEM_BULLET_BAG:
-                randoGetItem = Randomizer_GetItemFromKnownCheck(RC_LW_TARGET_IN_WOODS, GI_BULLET_BAG_50);
+                randoGetItem = (CVarGetInteger("gEnhancement.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LW_TARGET_IN_WOODS))
+                                ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_LW_TARGET_IN_WOODS, GI_BULLET_BAG_50);
```
```suggestion
                                ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_LW_TARGET_IN_WOODS, GI_BULLET_BAG_50);
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c:
```diff
@@ -540,8 +543,8 @@ void EnExItem_DrawHeartPiece(EnExItem* this, PlayState* play) {
-        GetItemEntry randoGetItem =
-            Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_SECOND_PRIZE, GI_HEART_PIECE);
+        GetItemEntry randoGetItem = (CVarGetInteger("gEnhancement.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_MARKET_BOMBCHU_BOWLING_SECOND_PRIZE))
+                                    ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_SECOND_PRIZE, GI_HEART_PIECE);
```
```suggestion
                                    ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_SECOND_PRIZE, GI_HEART_PIECE);
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_GirlA/z_en_girla.c:
```diff
@@ -1332,7 +1332,8 @@ void EnGirlA_Draw(Actor* thisx, PlayState* play) {
     if (this->actor.params == SI_RANDOMIZED_ITEM) {
         ShopItemIdentity shopItemIdentity = Randomizer_IdentifyShopItem(play->sceneNum, this->randoSlotIndex);
-        GetItemEntry getItemEntry = Randomizer_GetItemFromKnownCheckWithoutObtainabilityCheck(shopItemIdentity.randomizerCheck, shopItemIdentity.ogItemId);
+        GetItemEntry getItemEntry = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && this->actor.params == SI_RANDOMIZED_ITEM) ? GetItemMystery : 
```
```suggestion
        GetItemEntry getItemEntry = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && this->actor.params == SI_RANDOMIZED_ITEM) ? GetItemMystery() : 
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Si/z_en_si.c:
```diff
@@ -213,7 +213,8 @@ void EnSi_Draw(Actor* thisx, PlayState* play) {
         } else {
-            getItem = Randomizer_GetItemFromActor(this->actor.id, play->sceneNum, this->actor.params, GI_SKULL_TOKEN);
+            RandomizerCheck check = Randomizer_GetCheckFromActor(this->actor.id, play->sceneNum, this->actor.params);
+            getItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(check)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(check, GI_SKULL_TOKEN);
```
```suggestion
            getItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(check)) ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(check, GI_SKULL_TOKEN);
```

briaguya0 commented on soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c:
```diff
@@ -249,9 +250,9 @@ void ItemEtcetera_Draw(Actor* thisx, PlayState* play) {
         GetItemEntry randoGetItem = (GetItemEntry)GET_ITEM_NONE;
         if (type == ITEM_ETC_ARROW_FIRE) {
-            randoGetItem = Randomizer_GetItemFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE);
+            randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LH_SUN)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE);
```
```suggestion
            randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LH_SUN)) ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE);
```

briaguya0 commented on soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c:
```diff
@@ -249,9 +250,9 @@ void ItemEtcetera_Draw(Actor* thisx, PlayState* play) {
+            randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LH_SUN)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE);
         } else if (type == ITEM_ETC_LETTER) {
-            randoGetItem = Randomizer_GetItemFromKnownCheck(RC_LH_UNDERWATER_ITEM, GI_LETTER_RUTO);
+            randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LH_UNDERWATER_ITEM)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_LH_UNDERWATER_ITEM, GI_LETTER_RUTO);
```
```suggestion
            randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_LH_UNDERWATER_ITEM)) ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_LH_UNDERWATER_ITEM, GI_LETTER_RUTO);
```

briaguya0 commented on soh/src/overlays/actors/ovl_Item_Ocarina/z_item_ocarina.c:
```diff
@@ -215,7 +215,7 @@ void ItemOcarina_Draw(Actor* thisx, PlayState* play) {
 
     if (IS_RANDO) {
-        GetItemEntry randoGetItem = Randomizer_GetItemFromKnownCheck(RC_HF_OCARINA_OF_TIME_ITEM, GI_OCARINA_OOT);
+        GetItemEntry randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_HF_OCARINA_OF_TIME_ITEM)) ? GetItemMystery : Randomizer_GetItemFromKnownCheck(RC_HF_OCARINA_OF_TIME_ITEM, GI_OCARINA_OOT);
```
```suggestion
        GetItemEntry randoGetItem = (CVarGetInteger("gRandoEnhancements.MysteriousShuffle", 0) && Randomizer_IsCheckShuffled(RC_HF_OCARINA_OF_TIME_ITEM)) ? GetItemMystery() : Randomizer_GetItemFromKnownCheck(RC_HF_OCARINA_OF_TIME_ITEM, GI_OCARINA_OOT);
```

Malkierian committed:
Cross-platform building edits. If anyone has a better idea of how to handle this fake GIE, I'm all ears.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into freestanding-mystery

Malkierian committed:
Update to CVar macros.

Pepper0ni commented:
> Also utilizes GetMerchantMessage's mysterious option directly based on the setting, so is applied to all item names in purchasable situations.

Does this apply to only shops or to Scrubs and Solo merchants too?

Malkierian commented:
> > Also utilizes GetMerchantMessage's mysterious option directly based on the setting, so is applied to all item names in purchasable situations.
> 
> Does this apply to only shops or to Scrubs and Solo merchants too?

I believe I made it apply to scrubs and merchants as well. And even the bean salesman, I think.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into freestanding-mystery

Pepper0ni commented:
I don't like the idea of having merchant and scrub "mysterious item" settings being both in seed generation and in an enhancement. It's redundant and will end in confusion when people set their seed to do 1 thing and the game does another because of an enhancement.

Shops Make sense to be tied to the visual so I don't especially mind those, but it does beg the question as to why 1 set of mysteriousness is in enhancements and 1 is in seed generation. 

I don't consider this blocking as it's pointed at develop, but for v3 I would want something consistent, especially as scrub item names pose me problems I don't know how to properly deal with just yet.

Malkierian commented:
The only counter I could give to that is spoiler sharing, where a person might get a spoiler with merchant and scrub hints there, but still want to obfuscate them for their own run, but I don't know how often that would happen.

Pepper0ni commented:
If that is desirable, then all the mysterious settings should be in enhancements, which is reasonable as they don't affect seed generation in any way.

Pepper0ni commented:
![image](https://github.com/HarbourMasters/Shipwright/assets/93387759/6ed1009b-3868-47f0-92bb-3fe908fb0a44)
colours for mystery items do not seem to be working.

Pepper0ni commented:
Also, whenever I pick up a mysterious item, it's a heart piece, regardless of what the actual item is, unless it's from a shop

Malkierian commented:
*Sigh* How many times am I going to have to actually fix this before it finally gets merged...

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into freestanding-mystery

Malkierian committed:
Fix freestanding item gives.
Fix mysterious item model colors.

Malkierian commented:
Alright, I fixed the colors and the heart piece freestanding item gives. From what I could tell, gold skulltulas, small keys, heart containers, and the OoT were still working. If you come across any specific source that isn't working, let me know.

Malkierian commented:
You'll need a full rebuild, including cmake, after pulling these changes, I updated to latest develop.

Pepper0ni commented:
Is it intended for bombchu bowling to be unaffected?

Malkierian commented:
No, that is not intended. I'll look into it.

Malkierian committed:
Fix bombchu bowling mystery.

Malkierian commented:
Fixed.

Pepper0ni commented:
game crashes when loading 3rd round of bombchu bowling (first non-check round), going to set up a build now to get a call stack

Malkierian commented:
Couldn't replicate in debug mode. Might need more specific steps.

Malkierian committed:
Remove bowling bomchus check obfuscation (unused, removed in v3).

Pepper0ni commented on soh/soh/Enhancements/cosmetics/CosmeticsEditor.cpp:
```diff
@@ -322,6 +322,7 @@ static std::map<std::string, CosmeticOption> cosmeticOptions = {
     COSMETIC_OPTION("World.Moon",                    "Moon",                 COSMETICS_GROUP_WORLD,        ImVec4(240, 255, 180, 255), false, true, true),
     COSMETIC_OPTION("World.GossipStone",             "Gossip Stone",         COSMETICS_GROUP_WORLD,        ImVec4(200, 200, 200, 255), false, true, true),
     COSMETIC_OPTION("World.RedIce",                  "Red Ice",              COSMETICS_GROUP_WORLD,        ImVec4(255,   0,   0, 255), false, true, false),
+    COSMETIC_OPTION("World.MysteryItem",             "Mystery Item",         COSMETICS_GROUP_WORLD,        ImVec4(0,    60, 100, 255), false, true, false),
```
I'm not sure if world is the best category for this, as I looked all the way through items looking for this first, but there's no good section in items for it either and this is subjective so should get more feedback, so I will not consider this blocking.

Pepper0ni reviewed (approved):
The other small issue I noticed is that having mysterious items on causes the check tracker to hide the prices of things in shops as well as the item identity, which is inconvenient, but that's a small thing I'm willing to kick to the v3 implementation given that large chunks of related code is being redone there anyway. 

otherwise, LGTM.

inspectredc commented on soh/soh/Enhancements/randomizer/draw.cpp:
```diff
@@ -275,5 +275,28 @@ extern "C" void Randomizer_DrawTriforcePieceGI(PlayState* play, GetItemEntry get
+
+    Gfx_SetupDL_25Xlu(play->state.gfxCtx);
+
+    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(play->state.gfxCtx, (char*)__FILE__, __LINE__), G_MTX_MODELVIEW | G_MTX_LOAD);
```
```suggestion
    gSPMatrix(POLY_XLU_DISP++, MATRIX_NEWMTX(play->state.gfxCtx), G_MTX_MODELVIEW | G_MTX_LOAD);
```

Malkierian replied:
Following review on Discord, this will require a change in macros.h that is probably outside the purview of this PR, so I'll go ahead and mark this as resolved.

inspectredc commented on soh/soh/Enhancements/randomizer/draw.cpp:
```diff
@@ -275,5 +275,28 @@ extern "C" void Randomizer_DrawTriforcePieceGI(PlayState* play, GetItemEntry get
+    gDPSetGrayscaleColor(POLY_XLU_DISP++, color.r, color.g, color.b, 255);
+    gSPGrayscale(POLY_XLU_DISP++, true);
+
+    gSPDisplayList(POLY_XLU_DISP++, (Gfx*)gMysteryItemDL);
```
Do we need the cast here?

Malkierian replied:
I assume so, I was just following the pattern elsewhere, like with TriforceCompletedDL.

inspectredc reviewed (approved):
LGTM! just a couple unimportant things in the draw section
